From 1def2965d8da8cc74ab0036d7f8d59e81e676cad Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Sebastian=20Dr=C3=B6ge?= <sebastian@centricular.com>
Date: Fri, 27 Sep 2024 15:50:54 +0300
Subject: [PATCH] qtdemux: Check sizes of stsc/stco/stts before trying to merge
 entries

Thanks to Antonio Morales for finding and reporting the issue.

Fixes GHSL-2024-246
Fixes https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/3854

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/8060>
---
 .../gst-plugins-good/gst/isomp4/qtdemux.c     | 22 +++++++++++++++++++
 1 file changed, 22 insertions(+)

--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -10029,6 +10029,21 @@ qtdemux_merge_sample_table (GstQTDemux *
     return;
   }
 
+  if (gst_byte_reader_get_remaining (&stream->stts) < 8) {
+    GST_DEBUG_OBJECT (qtdemux, "Too small stts");
+    return;
+  }
+
+  if (stream->stco.size < 8) {
+    GST_DEBUG_OBJECT (qtdemux, "Too small stco");
+    return;
+  }
+
+  if (stream->n_samples_per_chunk == 0) {
+    GST_DEBUG_OBJECT (qtdemux, "No samples per chunk");
+    return;
+  }
+
   /* Parse the stts to get the sample duration and number of samples */
   gst_byte_reader_skip_unchecked (&stream->stts, 4);
   stts_duration = gst_byte_reader_get_uint32_be_unchecked (&stream->stts);
@@ -10040,6 +10055,13 @@ qtdemux_merge_sample_table (GstQTDemux *
   GST_DEBUG_OBJECT (qtdemux, "sample_duration %d, num_chunks %u", stts_duration,
       num_chunks);
 
+  if (gst_byte_reader_get_remaining (&stream->stsc) <
+      stream->n_samples_per_chunk * 3 * 4 +
+      (stream->n_samples_per_chunk - 1) * 4) {
+    GST_DEBUG_OBJECT (qtdemux, "Too small stsc");
+    return;
+  }
+
   /* Now parse stsc, convert chunks into single samples and generate a
    * new stsc, stts and stsz from this information */
   gst_byte_writer_init (&stsc);
